src/gil.rs

```diff
@@ -46,66 +46,125 @@ pub(crate) fn gil_is_acquired() -> bool {
-                        ffi::PyGILState_Ensure();
-                        ffi::Py_Finalize();
-                    }
-                }
-            }
-            libc::atexit(finalize);
-
-            // > Changed in version 3.7: This function is now called by Py_Initialize(), so you don’t have
-            // > to call it yourself anymore.
-            #[cfg(not(Py_3_7))]
-            if ffi::PyEval_ThreadsInitialized() == 0 {
-                ffi::PyEval_InitThreads();
-            }
+            #[cfg(not(Py_3_7))] // Called by Py_InitializeEx in Python 3.7 and up.
+            ffi::PyEval_InitThreads();
 
-            // Py_InitializeEx() will acquire the GIL, but we don't want to hold it at this point
-            // (it's not acquired in the other code paths)
-            // So immediately release the GIL:
-            let _thread_state = ffi::PyEval_SaveThread();
-            // Note that the PyThreadState returned by PyEval_SaveThread is also held in TLS by the Python runtime,
-            // and will be restored by PyGILState_Ensure.
+            // Release the GIL.
+            ffi::PyEval_SaveThread();
         }
     });
 }
 
+/// Executes the provided closure with an embedded Python interpreter.
+///
+/// This function intializes the Python interpreter, executes the provided closure, and then
+/// finalizes the Python interpreter.
+///
+/// After execution all Python resources are cleaned up, and no further Python APIs can be called.
+/// Because many Python modules implemented in C do not support multiple Python interpreters in a
+/// single process, it is not safe to call this function more than once. (Many such modules will not
+/// initialize correctly on the second run.)
+///
+/// # Availability
+/// This function is only available when linking against Python distributions that contain a
+/// shared library.
+///
+/// This function is not available on PyPy.
+///
+/// # Panics
+/// - If the Python interpreter is already initalized before calling this function.
+///
+/// # Safety
+/// - This function should only ever be called once per process (usually as part of the `main`
+///   function). It is also not thread-safe.
+/// - No Python APIs can be used after this function has finished executing.
+/// - The return value of the closure must not contain any Python value, _including_ `PyResult`.
+///
+/// # Example
+/// ```rust
+/// use pyo3::prelude::*;
+///
+/// # #[allow(clippy::needless_doctest_main)]
+/// fn main() {
+///     unsafe {
+///         pyo3::with_embedded_python_interpreter(|py| {
+///             py.run("print('Hello World')", None, None)
+///         });
+///     }
+/// }
+/// ```
+#[cfg(all(Py_SHARED, not(PyPy)))]
+pub unsafe fn with_embedded_python_interpreter<F, R>(f: F) -> R
+where
+    F: for<'p> FnOnce(Python<'p>) -> R,
+{
+    assert_eq!(
+        ffi::Py_IsInitialized(),
+        0,
+        "called `with_embedded_python_interpreter` but a Python interpreter is already running."
+    );
```

Comment:
If we force some error type (e.g., `PyErr`), we can return an `Err` here. It's an optional choice, though.

Reply:
I think `PyErr` doesn't work here because then the user will have a `PyErr` outside of the scope of a Python interpreter!

The hazards of `unsafe` 😅